### DISCUSSION TITLE
Eclipse Platform - Generating Stronger DSA & RSA Keys

### DISCUSSION THREAD

**Discussion by axmanalad** [eclipse-ide/contributing#2] [Show and tell]
As October came to a close, my micro-internship opportunity with [CodeDay](https://www.linkedin.com/company/codeday-org/) Labs ended as well. I was given the opportunity by [Computing Talent Initiative](https://www.linkedin.com/company/computing-talent-initiative/) (CTI) and CodeDay Labs to collaborate with a team and work in an open source contribution for [Eclipse Foundation](https://www.linkedin.com/company/eclipse-foundation/) project. I would like to share my insights on how my experience felt like with the team.

Firstly, I would like to thank [Sathya Narayanan](https://www.linkedin.com/in/sathyanarayanan6/) and [Utsab Saha](https://www.linkedin.com/in/sahautsab/) for prepping my advancement to tackle my first open source contribution when participating in the CTI program. Moreover, I also would like to place my gratitude for my teammates, [Shawn Lin](https://www.linkedin.com/in/shawn-lin-349449240/) and [Chelsy L.](https://www.linkedin.com/in/chelsy-l-68b613197/) for supporting and ensuring small but powerful tips into dedication for this project, including our technical mentor [Polina Keselman](https://www.linkedin.com/in/polina-keselman-7b62a81/) for her valuable wisdom.

This transformative journey has only begun, and I cannot wait to find more opportunities. To understand the how my contribution impacted the Eclipse Platform, please take a deep dive into my inspiring article.

![image](https://github.com/user-attachments/assets/da85afcd-4643-423e-b89f-3be446a6d687)
### Introduction
For one month, I was given the opportunity by [CodeDay Labs](https://labs.codeday.org/) and [Computing Talent Initiative (CTI) Accelerate](https://computingtalentinitiative.org/accelerate/) to contribute to one of the largest Java IDEs, the [Eclipse Platform](https://github.com/eclipse-platform/eclipse.platform). I was assigned to a team of three with a mentor to solve an [issue](https://github.com/eclipse-platform/eclipse.platform/issues/1464) for this project. From struggling to create the development environment, to searching through the codebase, to brief communications, I saw that our team as a whole was new to contributing; I found that our team revolved on complexities that we needed to embrace towards. In this article, I will present the progression of how our team was able to advance from reading a codebase to compiling and presenting a solution to the project, to gaining another follow-up opportunity. Not only has this project introduced us to the world of open source contribution but it also reflected a learning experience into the steps of programming real world solutions.

### The Codebase
Eclipse-Platform is an open source project which is the base framework for the integrated development environment _Eclipse_. It is the second most popular IDE for Java and was the most popular until 2016. _Eclipse_ is also a free open-source IDE that any programmer can use through multiple operating systems (Windows, Mac, and Linux) and extensible plug-in for customization. 

![image](https://github.com/user-attachments/assets/bb4c3f9b-cadb-4cf9-867c-a355d42cc569)
_First time opening up Eclipse IDE greets with a “Welcome” page_

For Java developers, _Eclipse_ is an important IDE for developing computer applications. This IDE is utilized for computer programming, mainly to write programs or create software with Java, but it can also be written in other programming languages such as C and Python. As the base framework, it holds all the functionalities and services it needs to operate _Eclipse_ for development purposes. These services perform every necessary function a developer would need to create computer applications and optimize their workflow.

For example,  a Java developer can develop a mobile application by writing Java code in their coding environment, finding code suggestions by Intellisense, debugging in real-time, and pushing code through a remote server like GitHub inside of _Eclipse_.
Since _Eclipse_ has been around for more than twenty years and is currently the second most popular IDE for Java development, there are millions of users who still use Eclipse IDE to develop applications with Java.

![image](https://github.com/user-attachments/assets/82f18449-3fac-469b-af26-2f9b95e4908f)
_A Java project showcase of “HelloWorld” utilizing the Eclipse IDE._

In the open source of the Eclipse Platform, committers join in [GitHub](https://github.com/eclipse-platform) to contribute to the project. There are 22 repositories that are based on several Java libraries like **Ant** and **Maven**. These libraries would then be utilized into repositories as modules that link together to perform operations for other Eclipse repositories such as build tools, the Standard Widget Tool (SWT), etc.

![image](https://github.com/user-attachments/assets/cc13b8d5-0361-4ac1-8461-14f8e0d32420)
![image](https://github.com/user-attachments/assets/48936b39-86df-4190-99b5-7e4ac28c91ff)
_Showcase of Eclipse Platform’s repositories of open contributions._

### Tech Stack
Before heading into the issue, there are a few tech stacks that should be interpreted.

![image](https://github.com/user-attachments/assets/938b5b54-8f5f-4d33-b792-45f95e32cfda)
_Important tech stacks of the Eclipse Platform._

**Java Programming Language** - This is the core language that Eclipse primarily utilizes to implement the entire IDE. According to the [master repository of Eclipse Platform](https://github.com/eclipse-platform/eclipse.platform), 96.7% of its repository is made with Java. There are also numerous libraries and modules that the Eclipse Platform integrated with purely Java that it honestly works as the back-end of its program as well.
**Workbench, Standard Widget Tool, JFace** - All of these frameworks/toolkits  incorporate some user interaction. For example, Workbench is responsible for the functionality of the IDE and provides visual options like toolbars and editors. Standard Widget Tool is responsible for supplying graphical interfaces that ensure quality appearance. JFace is responsible for simplifying complex interfaces by handling dialog boxes, wizards, etc.
**Equinox** - Although the Eclipse Platform does not provide a natural back-end framework, Equinox is the closest by unloading and loading plugins, bundles, and services; allows developers to run applications as bundles.
**Java Secure Channel** - This is a Java library that is responsible for creating secure/SSH connections between two computers; essential for key management such as DSA and RSA key pairs when authenticating SSH for remote connections like Git. This is also very important for the issue.

### **About the Issue**
The codebase handled weak keys, specifically **RSA** (Rivest-Shamir-Aldeman) cryptography. The issue that we tackled for the Eclipse-Platform is: When generating an RSA key, it is always defaulted to RSA-1024, a weak encryption that is not as secure as RSA-2048 or RSA-4096.

![image](https://github.com/user-attachments/assets/cb1dfb52-9936-47fc-b4e8-defb3cb7d09d)
_Locating the Preferences Page for accessing the key generator._

Normally when you open the Eclipse IDE, the application will open your workspace. To generate an RSA-key, you will navigate through Window => Preferences. Afterwards, you can either search up “SSH2” or navigate through General => Network Connections => SSH2. Go into Key Management and there you will find two key generators.

The RSA key generates a random RSA-1024 key to save locally when pressing the “Generate RSA Key..” button.

![image](https://github.com/user-attachments/assets/14323751-473d-4d27-b7db-6a5c45562e29)
_Showcase of the Key Management, where the RSA key becomes generated._

Meanwhile, what the issue declares to be satisfied is that the key generator should have a larger key size, such that RSA-2048 or RSA-4096 could be generated instead for stronger encryption. A stronger RSA encryption will allow a more secure transmission of sensitive data and protects data exchange between the server and client. This is especially important for open-source projects of many developers that may want to retrieve or transmit data between their side and the server they are working on. For a better visual of how the keys are generated, a system diagram is provided below.

### Flow Control of Eclipse’s Key Generation
![image](https://github.com/user-attachments/assets/9a970358-f752-4f76-8e48-7bb67916cc35)
_A simple system diagram on how Eclipse handles key generations._

### Digging Into the Issue
Relevant to the issue of generating an RSA key, I had to identify key locations to where the codebase is generating an RSA key by pressing a button. Luckily, the initial comment of the [issue](https://github.com/eclipse-platform/eclipse.platform/issues/1464) provided a hint that the button was in the “Preferences” page. I made a few comments regarding this location to understand how the RSA key gets generated.

![image](https://github.com/user-attachments/assets/713a5c12-d9ea-4c46-9765-9e4b3a05a2ee)
_Location of the issue in the codebase of PreferencePage.java file._

Another hint proved that this was the location of the issue by exploring the “KeyPair.class”, where a method called “genKeyPair” has a default key size of 1024 bits.

![image](https://github.com/user-attachments/assets/b676e530-b96d-4339-bc7b-4429ab397b3a)
_Image of genKeyPair function in KeyPair.class_

I had to explore through other classes to understand how each of the new objects work, particularly “KeyPairRSA(jsch)”, which generates the key pair to be in RSA format. From here, I noticed that one of the tech stacks for this issue was the **Java Secure Channel** (**JSch**), a Java implementation meant for providing secure connection between clients. Once I figured out how the codebase worked for the particular functionality, I proceeded to start implementing a solution.

### Implementing a Solution
The first thing I quickly checked was verifying whether the location of the issue was correct. This was verified by placing print statements for the console to return feedback if the statement was called inside of the function. Once I figured out the statement was called by testing the key generator button, I started to modify the codebase with a simple if and else statement.

![image](https://github.com/user-attachments/assets/8caf9929-e8c7-47eb-9666-cd9578a3ff95)
_Before Code Modification_

![image](https://github.com/user-attachments/assets/2e2c44b9-cc29-46e4-b7c9-b17e1fa61965)
_After Code Modification_

My first solution was to fixate the RSA key size to 2048 bits, acknowledging the issue’s recommendation. This was implemented by changing both the “genKeyPair” function and the kpairComment to their respective bit size. The **DSA** (Digital Signature Algorithm) key size was kept to 1024 bits.

One week later, a few knowledgeable people and the project manager [answered](https://github.com/eclipse-platform/eclipse.platform/issues/1464#issuecomment-2439600327) my questions that allowed me to adjust my solution. This adjustment upgraded from 2048 bits to 4096 bits as the key size for RSA. DSA would also be formatted to provide 3072 bits instead of 1024 bits. Notably, DSA-4096 would give an error instead of DSA-3072 based on CPU runtime performance; the larger the key size, the more disk storage is utilized, which increases CPU runtime. By this error, I comprehended the idea that DSA-4096 consumes too much disk storage to compute a DSA key with a size of 4096 bits. Despite a decrease in performance, it is better to have a larger key size for more security.

### Challenges Before Pull Requesting and Committing the Solution
One challenge prior to searching through the codebase was integrating a working environment. On the first day of attempting to get the workspace, the [Contributing Page](https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md) supplied the instructions, but I had trouble running the workspace as well as connecting to my GitHub forked repository.

The next day with a clear mind, I discovered my faults such as using IntelliJ IDEA as my IDE, attempting to use Visual Studio Code/GitHub’s codespaces, and not installing all packages. I was able to fix my problem by carefully reading the installation instructions as well as using Git Bash and Linux commands to create separate remotes linking to both my forked repository and the main repository. I would learn that following any contributing instructions is essential to open source contributions.

Another challenge involved vague instructions in the issue, such that I do not know whether the issue wants multiple options of generating different key sizes or whether the key size should only affect the RSA key and not the DSA key. The problem with editing the key generator function in the PreferencePage.java file is that both RSA and DSA keys would be affected if we were to add an argument to the key size parameter. I would have to ask about these [questions](https://github.com/eclipse-platform/eclipse.platform/issues/1464#issuecomment-2421088000) in the issue discussion.

![image](https://github.com/user-attachments/assets/baa22c09-0e63-4840-b32c-fe26f838b154)
_Questions that I asked about the issue (Date: 10/17/2024)_

Ever since I asked however, I have never received any feedback from the project managers. After almost a week, my team and mentor joined in a meeting to discuss our progress, and my mentor responded to just continue with the solution despite finding answers from the management team. I would submit my [first pull request](https://github.com/eclipse-platform/eclipse.platform/pull/1596) to the Eclipse Platform project, which was accidentally merged by the project manager without review. However, the project manager was able to answer my questions two days later. With my new solution, I made a new pull request.

### The Pull Request Review Phase
The [new pull request](https://github.com/eclipse-platform/eclipse.platform/pull/1600) was made after refactoring the code based on the comment from the project manager. However, the problem with my previous solution involved many issues which were merge conflicts, failure on a few test cases, and messy commits. When the project manager noticed this, he made a [feedback](https://github.com/eclipse-platform/eclipse.platform/pull/1600#pullrequestreview-2400207287) mentioning about organization and compiling, including the usage of two Linux commands I had less experience with: _rebase_ and _squash_.

Before I tackled Git's rebase and squash, I had issues compiling the workspace to run the program, resulting in making the “Revert merge changes” commit. I had the practice of updating the workspace every time there is a new change to sync the fork repository and merge those changes into my local development environment. However, what I never noticed was the other repositories. Remember when I mentioned at the beginning of this blog about the project having 22 repositories? Apparently, these repositories also have to be updated to the latest changes locally. This was resolved according to the project manager’s comment by rerunning the _Oomph_ Eclipse installer, which is the setup needed to create the development environment for contributing. This grabbed the latest changes easily without merging all repositories. I would realize later on that there was a better alternative.

![image](https://github.com/user-attachments/assets/518530be-2dca-4823-a8ba-1b5e93bef94d)
_Image of Eclipse Installer made by Oomph_

After resolving that issue, I was less familiar with the concepts of rebasing and squashing, which required me to do a bit of research through Linux commands. I researched mostly from _StackOverflow_ threads, absorbing more information such as how to remove a commit in the middle of a commit history and combining commits into a single commit. By the time I rebased and squashed the commits into a single clean commit, I forced-pushed the changes into the pull request. All the test cases have passed flawlessly and I followed up with an appreciative comment for his feedback. 

![image](https://github.com/user-attachments/assets/3b672e29-9f5a-4522-b4b1-f35d62238b12)
![image](https://github.com/user-attachments/assets/d73628db-f750-4570-ab2a-e3aa1ea04844)
_Eclipse’s Test Cases after making the changes._

Afterwards, the project manager requested [an organization of code](https://github.com/eclipse-platform/eclipse.platform/pull/1600#pullrequestreview-2411927494), such that it simplifies the readability of the implementation. Once I cleaned up the code with his request, everything was satisfactory to earn a merge as the final solution.

### Final Solution
In summary, the [code changes](https://github.com/eclipse-platform/eclipse.platform/pull/1600/files) reflect on a key size difference, which affects the output of key generation. By having a larger key size, it is more secure for Java developers to continue programming their projects while using the Eclipse IDE.

![image](https://github.com/user-attachments/assets/3fd09ad5-ca50-48a5-a2d5-8a0cab032591)
_Before Code Changes (RSA-1024 Key Sample)_

![image](https://github.com/user-attachments/assets/7e57df4a-5719-4e79-8660-3c04c69fe464)
_After Code Changes (RSA-4096 Key Sample)_

It is an extreme honor to work with my partners along with my mentor to get through this journey. This is an opportunity I would truly never forget to experience the growth of software engineering, and I know my path only starts from here. I would like to thank a project manager of the Eclipse Platform, Hannes Wellmann. He has assisted and suggested an abundant amount of my progression towards developing the solution to the issue. As of November 4, 2024, I am in the process of making another contribution by a follow-up addition which allows the user to select the key size of the generated key.

_This discussion was initially posted on [LinkedIn](https://www.linkedin.com/posts/axmanalad_as-october-came-to-a-close-my-micro-internship-activity-7259595356173213698-8sbX?utm_source=share&utm_medium=member_desktop)._

**Comment by tfroment** [comment]
Thank you @axmanalad for sharing this great feedback with us here!
